### ISSUE TITLE
Denoising Diffusion Probabilistic Models

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#154]
## TL; DR

- Reparameterize diffusion model variational bound objective so that it resembles that of denoising score matching.

## Motivation

- Diffusion models are straightforward to define and efficient to train, but sample quality has lagged behind state-of-the-art generative models such as GANs.

## Prior Works

- Diffusion Model (Sohl-Dickstein, 2015)
- Variational Autoencoders
- Autoregressive Models (PixelCNN, PixelSnail)

### Objective

- Variational Bound on NLL (identical to that of VAEs)

![image](https://user-images.githubusercontent.com/46308407/152931982-ae5038c2-8828-40b8-843f-f6fdd4ead6e8.png)

- Simple modification using Bayes formula and conditional probabilities

![image](https://user-images.githubusercontent.com/46308407/152932007-3f84535c-e8d8-4537-a0d9-4b80159a8b77.png)

- We see that the objective consists of three components: (1) Forward process should completely *destroy* the image and transform it to Gaussian (2) Intermediate learned denoising steps should resemble that of closed-form denoising conditioned on GT image (3) Final output from learned denoising steps should be likely

- Term (1) can be ignored since we use constants for forward variance \beta (linearly scaled with adequate magnitude)
- Term (2) is of special interest, as it affects the training procedure the most

![image](https://user-images.githubusercontent.com/46308407/152932800-801f0a78-42cc-4444-9509-7cdb0fffd52e.png)

- We have several **design choices** to make: forward process variance, model architecture, Gaussian parameterization of reverse process
- Parameterizing our predicted Gaussian mean appropriately, we obtain an objective similar to that of denoising autoencoders:

![image](https://user-images.githubusercontent.com/46308407/152932831-6e71f11b-b815-4749-b9f9-a6cddd0f87eb.png)
![image](https://user-images.githubusercontent.com/46308407/152932866-d0da5de2-c6f3-4d3f-ba13-8db212fb163e.png)

- For Term (3), we use independent discrete decoder (simple form)
![image](https://user-images.githubusercontent.com/46308407/152932951-53c2de01-4a0c-4bce-b8c6-3ec5502ce637.png)

- In actual training, the authors found using a simplified denoising objective beneficial for sample qualitiy
![image](https://user-images.githubusercontent.com/46308407/152933151-e024ed9a-ba1c-407b-8d50-7ae0fcc6f31b.png)

- Intuitively, discarding the weights (hence weighted variational bound) down-weights loss terms with small t (stage where the input is almost identical to the GT image) and relatively up-weights terms with large t (where there is much noise hence difficult).
- Authors sample t from U[1,T].

### Implementation

- For backward process, use U-Net backbone similar to PixelCNN++.
- Forward covariance scale \beta is linearly scaled from 0.0001 to 0.01.

![image](https://user-images.githubusercontent.com/46308407/152933609-bcabdb06-104d-4fe7-9223-92aed687e2d9.png)

- Note that complete sampling (x_T to x_0) resembles Langevin dynamics with e_\theta as *learned gradient of data log density*.

## Experiments

![image](https://user-images.githubusercontent.com/46308407/152933784-cbd02907-2c99-4a8f-b163-88b54559942b.png)
![image](https://user-images.githubusercontent.com/46308407/152933847-bffa2a65-d73b-42dd-b182-4f5b462273d4.png)
![image](https://user-images.githubusercontent.com/46308407/152933861-9b834ec0-45e1-40f3-bec3-2cef968b76b5.png)
![image](https://user-images.githubusercontent.com/46308407/152933891-f5f86121-57de-4fa7-8365-c2d6c8fffda9.png)
![image](https://user-images.githubusercontent.com/46308407/152933873-fa6fe538-84ee-4b21-9ed8-67b955990479.png)

### Strength

- Present potential for a new family of generative models (i.e., diffusion probabilistic models)
- Excellent theoretical narrative with connections to denoising score matching and Langevin dynamics. (score-based modeling)
- Nice engineering that matches the performance of SOTA generative models (such as StyleGAN2)

### Connection to VAE

- Identical Objective
- VAEs learn mapping from image space to embedding space that is regularized to resemble Gaussian (*the encoder*) while diffusion models employ rule-based transition (forward noising process).
- VAEs sample images from Gaussian in a feed-forward fashion, while diffusion models go through Markov chains, producing samples in iterative manner (hence slow).
- Diffusion models (*in my personal view*) has more freedom for design choices, as we can parameterize backward process parameters as we want (the authors opted for denoising)
- Both very elegant theoretically.

### Overall
- Excellent paper. Really impressive how they put motivation, theory and evaluation together in such a way that the paper is so smooth to read while maintaining persuasive tone.
